Batched dot run per subdirectory

Gabrielcarvfer opened the pull request:
I was building a project on Windows and noticed that Dot (graphviz) takes a really long time, while most doxygen processing goes reasonably fast. For the ns-3 project, the doxygen pre-processing, parsing, etc took 7 minutes while Dot processing took another 82 minutes using 16 threads.

That felt really wrong, so tested for a while and noticed just starting a single dot process takes up to 240 ms when Windows is under load. Since a new Dot process is launched for every .dot file, sometimes more than once to handle different output formats, it makes sense to take a long while to process everything.

Figured out DOT_MULTI_TARGETS exist, but it still isn't much faster than running everything in a single process, which took about 36 minutes. Not ideal either, but way better. Command used: `measure-command{dot -Tpng -O *.dot}`

Then I tried using CREATE_SUBDIRS and passing each subdirectory to a different Dot process. It ended up taking about 5 minutes. Command used: `measure-command{Get-ChildItem -Recurse -Directory | Foreach-Object -Parallel {cd $_.FullName; dot -Tpng -O *.dot}}`

This PR is a total hack, but I've implemented the second strategy and seems to work just fine. Same project now builds in ~16 minutes (doxygen+dot).

Gabrielcarvfer commented:
I probably took the stupid way that touches everything instead of simply renaming to remove the ".dot" from filename.dot.format_suffix, but it's just an idea.

doxygen commented:
@Gabrielcarvfer Interesting speedup that you observed. 

Do you know how much of this is related to the use of Windows (i.e. would there be a performance gain on Linux or MacOSX as well?), and do you have a virus scanner active and if so, do you know how much that impacts the performance?

Gabrielcarvfer commented:
> @Gabrielcarvfer Interesting speedup that you observed.
> 
> Do you know how much of this is related to the use of Windows (i.e. would there be a performance gain on Linux or MacOSX as well?), and do you have a virus scanner active and if so, do you know how much that impacts the performance?

It is definitely a Windows thing. And this was with Defender Real-time protection disabled, so it probably gets worse than that.

I've written a tiny script to test how long it takes to launch these processes.
<details><summary> Script</summary>

```
import argparse
import itertools
import os
import shutil
import subprocess
import sys
import time
from concurrent.futures import ThreadPoolExecutor

import matplotlib.pyplot as plt


def parse_argv(argv: list) -> argparse.Namespace:
    parser = argparse.ArgumentParser("Process time creation profiler")
    parser.add_argument("proc_args_to_profile", type=str, default=None)
    parser.add_argument("-n", "--num_procs", type=int, default=10000)
    parser.add_argument("--oversubscribe", action="store_true", default=False)
    args = parser.parse_args(argv[1:])
    return args


def experiment_procedure(experiment_id: int, program_to_profile: str) -> float:
    program_to_profile = program_to_profile.format(output_file=str(experiment_id))
    start = time.time_ns()
    ret = subprocess.run(program_to_profile.split(), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    return (time.time_ns() - start)/1e6


def run_experiment(program_to_profile: str, num_procs: int, num_proces: int) -> list:
    results = []
    with ThreadPoolExecutor(max_workers=num_proces) as pool:
        results = pool.map(experiment_procedure,
                           list(range(num_procs)),
                           itertools.repeat(program_to_profile)
                           )
        results = list(results)
    return results


def main(argv: list) -> int:
    args = parse_argv(argv)
    program = args.proc_args_to_profile.split()[0]
    program_path = shutil.which(program)
    if program_path is None:
        print("Program {program} was not found".format(program=program))
        exit(-1)

    measured_times = []
    lb_threads = 1+(os.cpu_count() if args.oversubscribe else 0)
    ub_threads = 1+(os.cpu_count()*2 if args.oversubscribe else os.cpu_count())
    for num_threads in range(lb_threads, ub_threads):
        start = time.time_ns()
        measured_times.append(
            run_experiment(
                args.proc_args_to_profile.replace(program, program_path),
                args.num_procs,
                num_threads
            )
        )
        print("Finished running measurements with {threads} threads.".format(threads=num_threads))
        print("Experiment time: {totTime} s. Sum of processes time: {time} s.".format(totTime=(time.time_ns()-start)/1e9,
                                                                                       time=sum(measured_times[-1])/1e3
                                                                                      ))
    plt.boxplot(
        measured_times,
        showmeans=True,
        labels=[str(i) for i in range(lb_threads, ub_threads)]
    )
    plt.title("Process time creation for different number of threads")
    plt.suptitle("Process args:'{proc}', Number of instances:'{nproc}'".format(proc=args.proc_args_to_profile,
                                                                               nproc=args.num_procs)
                                                                               )
    plt.ylabel("Time to launch and destroy process [ms]")
    plt.xlabel("Number of concurrent jobs [threads]")
    plt.tight_layout()
    plt.savefig("benchmark_process_results.png", dpi=300)
    return 0


if __name__ == "__main__":
    exit(main(sys.argv))

```
</details>

Plots are kind of misleading due to different Y-axis, so don't compare them side-by-side.
Orange tick = median
Green triangle = mean

<details><summary>Normal runs</summary>

* Ubuntu (median goes from 8 ms up to 10 ms)
![benchmark_process_results_ubuntu](https://user-images.githubusercontent.com/1678454/196041265-3ac9645a-d86d-419c-b09d-2deb51672be5.png)

* Windows 10 19044 (median goes from 14 ms up to 36 ms, Defender Real-Time Protection and behavior monitoring disabled)
![benchmark_process_results_windows_defenderoff](https://user-images.githubusercontent.com/1678454/196041799-838d897f-4976-4414-8de4-946946f2c6b3.png)

* Windows 10 19044 (median goes from 16 ms up to 47 ms, Defender Real-Time Protection and behavior monitoring enabled)
![benchmark_process_results_windows_defenderon](https://user-images.githubusercontent.com/1678454/196042418-6c4f521c-9462-4b1d-8a1b-e4e06ad6fda1.png)
</details>
<details><summary>Now "oversubscribing" (launching more processes than core threads available to run them)</summary>

Windows 10 19044 (median goes from 37 ms up to 73 ms, Defender Real-Time Protection and behavior monitoring disabled)
![benchmark_process_results_windows_defenderoff_oversubscribe](https://user-images.githubusercontent.com/1678454/196043823-d3f19f49-76f1-4304-87a2-f4bd525f1a14.png)


Windows 10 19044 (median goes from 50 ms up to 92 ms, Defender Real-Time Protection and behavior monitoring enabled)
![benchmark_process_results_windows_defenderon_oversubscribe](https://user-images.githubusercontent.com/1678454/196043811-e45c8f5f-251b-4274-8bc4-fb640b49b2de.png)

</details>

After testing, I believe I missed something. Still investigating.

Using a test.dot file
```
digraph G {
    one -> two;
    one -> three;
    one -> four;
    four -> five -> six;
    five -> one;
    six -> two;
}
```
I get the following output:
<details><summary>Output</summary>

```
$  python3 ./benchmarkProcessTimeCreation.py "dot -Tpng test.dot -o {output_file}"
Finished running measurements with 1 threads.
Experiment time: 151.179219 s. Sum of processes time: 150.76914919999962 s.
Finished running measurements with 2 threads.
Experiment time: 84.163069 s. Sum of processes time: 167.92437939999962 s.
Finished running measurements with 3 threads.
Experiment time: 60.5762176 s. Sum of processes time: 181.3378125000008 s.
Finished running measurements with 4 threads.
Experiment time: 47.4371849 s. Sum of processes time: 189.33227879999973 s.
Finished running measurements with 5 threads.
Experiment time: 38.7889991 s. Sum of processes time: 193.47421269999893 s.
Finished running measurements with 6 threads.
Experiment time: 34.1415998 s. Sum of processes time: 204.43076679999967 s.
Finished running measurements with 7 threads.
Experiment time: 31.2621731 s. Sum of processes time: 218.33634710000064 s.
Finished running measurements with 8 threads.
Experiment time: 28.4850091 s. Sum of processes time: 227.33824669999998 s.
Finished running measurements with 9 threads.
Experiment time: 26.4945616 s. Sum of processes time: 237.83164970000092 s.
Finished running measurements with 10 threads.
Experiment time: 25.1926108 s. Sum of processes time: 251.28324699999902 s.
Finished running measurements with 11 threads.
Experiment time: 24.3235777 s. Sum of processes time: 266.8940679999995 s.
Finished running measurements with 12 threads.
Experiment time: 23.8680912 s. Sum of processes time: 285.7712426999999 s.
Finished running measurements with 13 threads.
Experiment time: 23.5726652 s. Sum of processes time: 305.78775469999977 s.
Finished running measurements with 14 threads.
Experiment time: 23.3322343 s. Sum of processes time: 325.9033629000004 s.
Finished running measurements with 15 threads.
Experiment time: 23.1750562 s. Sum of processes time: 346.9748671000013 s.
Finished running measurements with 16 threads.
Experiment time: 23.145521 s. Sum of processes time: 369.4893681000004 s.

$ python3 ./benchmarkProcessTimeCreation.py "dot -Tpng test.dot -o {output_file}" --oversubscribe
Finished running measurements with 17 threads.
Experiment time: 22.8525207 s. Sum of processes time: 387.7279324999995 s.
Finished running measurements with 18 threads.
Experiment time: 23.1610485 s. Sum of processes time: 415.9718041000009 s.
Finished running measurements with 19 threads.
Experiment time: 23.7121131 s. Sum of processes time: 449.6790977000006 s.
Finished running measurements with 20 threads.
Experiment time: 23.0000463 s. Sum of processes time: 458.83599160000017 s.
Finished running measurements with 21 threads.
Experiment time: 22.4895292 s. Sum of processes time: 471.2831109000022 s.
Finished running measurements with 22 threads.
Experiment time: 23.2277996 s. Sum of processes time: 510.0205569999995 s.
Finished running measurements with 23 threads.
Experiment time: 22.9690541 s. Sum of processes time: 527.1832384000008 s.
Finished running measurements with 24 threads.
Experiment time: 23.108592 s. Sum of processes time: 553.3081685000021 s.
Finished running measurements with 25 threads.
Experiment time: 22.7835807 s. Sum of processes time: 568.0065223000003 s.
Finished running measurements with 26 threads.
Experiment time: 22.7395288 s. Sum of processes time: 589.8587014999988 s.
Finished running measurements with 27 threads.
Experiment time: 22.5735221 s. Sum of processes time: 607.8930167000016 s.
Finished running measurements with 28 threads.
Experiment time: 22.7245755 s. Sum of processes time: 634.8876631999977 s.
Finished running measurements with 29 threads.
Experiment time: 22.8661239 s. Sum of processes time: 661.0085386 s.
Finished running measurements with 30 threads.
Experiment time: 22.4511736 s. Sum of processes time: 671.8711473999988 s.
Finished running measurements with 31 threads.
Experiment time: 22.4565235 s. Sum of processes time: 694.4132699999958 s.
Finished running measurements with 32 threads.
Experiment time: 22.7125263 s. Sum of processes time: 724.8167211000016 s.
```
</details>

Which now makes me believe the primary issue isn't in launching dot, but how dot is being launched. 
Going to recheck the DotManager code.

Gabrielcarvfer commented:
Rewrote the threading part/queue of DotManager, but saw no improvement.

Thought it could be the Win32 variant of the Portable::System without console due to profiling with WPR/WPA.
However, only enabling the console on DotRunner didn't make any change.

Maybe it is just the way the calls are ordered in DotRunner::run.
